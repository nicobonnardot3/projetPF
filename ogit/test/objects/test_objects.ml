open Ogitlib
open Objects


(* on place le programme dans repo *)
let repo_root = "../../../../../repo"
let () = 
  Sys.chdir repo_root

(* on prepare le repertoire .ogit *)
let _ = Sys.command "rm -rf .ogit"
let _ = Sys.command "cp -r ../repo.origin/.ogit ."


(* test de la fonction hash sur des objets texte *)
let obj1 = Text "hello, world!"
let () = Format.printf "hash de obj1 : %s@." (Digest.to_hex (hash obj1))


(* test de la fonction hash sur des objets directory *)
let obj2 = Text "salut, le monde!"
let obj3 = Directory []
let obj4 = Directory [
  ("obj1", false, (hash obj1), obj1);
  ("obj2", false, (hash obj2), obj2);
  ("obj3", true, (hash obj3), obj3)
]
let () = Format.printf "hash de obj4 : %s@." (Digest.to_hex (hash obj4))


(* test de is_known *)
let hash_de_toto_txt = Digest.from_hex "1e423d8532170b358868e1af76688c4a"
let hash_bidon = Digest.from_hex "2f423d8532170b358868e1af76688c4a"

let () = Format.printf "is_known %s? %b@." (Digest.to_hex hash_de_toto_txt) (is_known hash_de_toto_txt) 
let () = Format.printf "is_known %s? %b@." (Digest.to_hex hash_bidon) (is_known hash_bidon)


(* test de store_object et read_text_object *)
let hash_obj4 = store_object obj4 
let text_obj4 = read_text_object hash_obj4
let () = Format.printf "@.CONTENU DU FICHIER REPRESENTANT L'OBJET obj4: @.%s@." text_obj4

(* on supprime obj4 *)
let _ = Sys.command ("rm -f .ogit/objects/" ^ (Digest.to_hex hash_obj4))

(* test de store_work_directory et read_directory_object *)
(* on suppose que le répertoire repo est dans l'état initial (celui fourni dans l'archive), à savoir 
.
├── foo
│   ├── bar
│   │   └── tata.txt
│   ├── titi.txt
│   └── toto.txt
└── toto.txt
avec les contenus des fichiers inchangés par rapport à ce qui était dans l'archive
On efface au début le contenu du répertoire .ogit/objects pour vérifier que store_work_directory fait bien son travail
Si le test échoue, il faudra restaurer ce répertoire en se basant sur l'archive fournie pour que les tests du début
passent à nouveau
*)
let () = Format.printf "@.EXECUTION DE LA COMMANDE SHELL tree@."
let _ = Sys.command "tree"
let hash_repo = store_work_directory ()
let () = Format.printf "@.hash_repo = %s@." (Digest.to_hex hash_repo)
let () = Format.printf "@.CONTENU DU FICHIER %s:@.%s@." (Digest.to_hex hash_repo) (read_text_object hash_repo)
let rec pp_object = function
| Text s -> Format.sprintf "Text(\"%s\")" s
| Directory(l) -> 
  l |> List.sort compare |> List.map (fun (s,d,h,obj) -> Format.sprintf "(\"%s\",%b,%s,%s)" s d (Digest.to_hex h) (pp_object obj))
  |> String.concat ";" 
  |> Format.sprintf "Directory[%s]"
let () = Format.printf "@.OBJET CORRESPONDANT AU REPO:@.%s@." (pp_object (read_directory_object hash_repo))


(* test de clean_work_directory et restore_work_directory *)
let () = Format.printf "@.EXECUTION DE clean_work_directory@."
let () = clean_work_directory ()
let () = Format.printf "@.EXECUTION DE LA COMMANDE SHELL tree -a@."
let _ = Sys.command "tree -a"
let () = Format.printf "@.EXECUTION DE restore_work_directory@."
let () = restore_work_directory (read_directory_object hash_repo)
let () = Format.printf "@.EXECUTION DE LA COMMANDE SHELL tree -a@."
let _ = Sys.command "tree -a"
let () = Format.printf "@.EXECUTION DE LA COMMANDE SHELL cat toto.txt@."
let _ = Sys.command "cat toto.txt"
