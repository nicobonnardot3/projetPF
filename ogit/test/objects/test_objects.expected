hash de obj1 : 3adbbad1791fbae3ec908894c4963870
hash de obj4 : eb300a7a2a303e6aee681bd0b0cda60a
is_known 1e423d8532170b358868e1af76688c4a? true
is_known 2f423d8532170b358868e1af76688c4a? false

CONTENU DU FICHIER REPRESENTANT L'OBJET obj4: 
obj1;t;3adbbad1791fbae3ec908894c4963870
obj2;t;84a21a9828c872e099f97fd9540ce398
obj3;d;d41d8cd98f00b204e9800998ecf8427e

EXECUTION DE LA COMMANDE SHELL tree
.
├── foo
│   ├── bar
│   │   └── tata.txt
│   ├── titi.txt
│   └── toto.txt
└── toto.txt

2 directories, 4 files

hash_repo = 9ef6d96530cb3791f0dfb16d157220b6

CONTENU DU FICHIER 9ef6d96530cb3791f0dfb16d157220b6:
toto.txt;t;1e423d8532170b358868e1af76688c4a
foo;d;c3ca6ad75c459ecccfdb10b421cbb489

OBJET CORRESPONDANT AU REPO:
Directory[("foo",true,c3ca6ad75c459ecccfdb10b421cbb489,Directory[("bar",true,87113b72a9d03cbfb8aea342455d710f,Directory[("tata.txt",false,a5b0164846456eb4c32115f57cfc3a38,Text("Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed non risus. 
"))]);("titi.txt",false,ec7e8e8ce16f3fe945d3c4c0955315d3,Text("Praesent blandit odio eu enim. Pellentesque sed dui ut augue blandit sodales. 
Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; 
Aliquam nibh. Mauris ac mauris sed pede pellentesque fermentum. 
Maecenas adipiscing ante non diam sodales hendrerit.
Ut velit mauris, egestas sed, gravida nec, ornare ut, mi. 
Aenean ut orci vel massa suscipit pulvinar. Nulla sollicitudin. 
Fusce varius, ligula non tempus aliquam, nunc turpis ullamcorper nibh, in tempus 
sapien eros vitae ligula. Pellentesque rhoncus nunc et augue. Integer id felis. 
"));("toto.txt",false,bb6d4c2870158394ea6b4c6ec804d61f,Text("Cras elementum ultrices diam. Maecenas ligula massa, varius a, semper congue, 
euismod non, mi. Proin porttitor, orci nec nonummy molestie, enim est eleifend mi, 
non fermentum diam nisl sit amet erat. Duis semper. 
Duis arcu massa, scelerisque vitae, consequat in, pretium a, enim. 
Pellentesque congue. Ut in risus volutpat libero pharetra tempor. 
Cras vestibulum bibendum augue. Praesent egestas leo in pede. 
"))]);("toto.txt",false,1e423d8532170b358868e1af76688c4a,Text("Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed non risus. 
Suspendisse lectus tortor, dignissim sit amet, adipiscing nec, ultricies sed, dolor. 
Cras elementum ultrices diam. Maecenas ligula massa, varius a, semper congue, 
euismod non, mi. Proin porttitor, orci nec nonummy molestie, enim est eleifend mi, 
non fermentum diam nisl sit amet erat. Duis semper. 
Duis arcu massa, scelerisque vitae, consequat in, pretium a, enim. 
Pellentesque congue. Ut in risus volutpat libero pharetra tempor. 
Cras vestibulum bibendum augue. Praesent egestas leo in pede. 
Praesent blandit odio eu enim. Pellentesque sed dui ut augue blandit sodales. 
Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; 
Aliquam nibh. Mauris ac mauris sed pede pellentesque fermentum. 
Maecenas adipiscing ante non diam sodales hendrerit.
Ut velit mauris, egestas sed, gravida nec, ornare ut, mi. 
Aenean ut orci vel massa suscipit pulvinar. Nulla sollicitudin. 
Fusce varius, ligula non tempus aliquam, nunc turpis ullamcorper nibh, in tempus 
sapien eros vitae ligula. Pellentesque rhoncus nunc et augue. Integer id felis. 
Curabitur aliquet pellentesque diam. Integer quis metus vitae elit lobortis egestas. 

Aliquam convallis sollicitudin purus. Praesent aliquam, enim at fermentum mollis, ligula massa adipiscing nisl, ac euismod nibh nisl eu lectus. Fusce vulputate sem at sapien. Vivamus leo. Aliquam euismod libero eu enim. Nulla nec felis sed leo placerat imperdiet. Aenean suscipit nulla in justo. Suspendisse cursus rutrum augue. Nulla tincidunt tincidunt mi. Curabitur iaculis, lorem vel rhoncus faucibus, felis magna fermentum augue, et ultricies lacus lorem varius purus. Curabitur eu amet."))]

EXECUTION DE clean_work_directory

EXECUTION DE LA COMMANDE SHELL tree -a
.
└── .ogit
    ├── HEAD
    ├── logs
    │   └── 934258186f418bbb65825ae6c2d2af31
    └── objects
        ├── 1e423d8532170b358868e1af76688c4a
        ├── 87113b72a9d03cbfb8aea342455d710f
        ├── 9ef6d96530cb3791f0dfb16d157220b6
        ├── a5b0164846456eb4c32115f57cfc3a38
        ├── bb6d4c2870158394ea6b4c6ec804d61f
        ├── c3ca6ad75c459ecccfdb10b421cbb489
        └── ec7e8e8ce16f3fe945d3c4c0955315d3

3 directories, 9 files

EXECUTION DE restore_work_directory

EXECUTION DE LA COMMANDE SHELL tree -a
.
├── .ogit
│   ├── HEAD
│   ├── logs
│   │   └── 934258186f418bbb65825ae6c2d2af31
│   └── objects
│       ├── 1e423d8532170b358868e1af76688c4a
│       ├── 87113b72a9d03cbfb8aea342455d710f
│       ├── 9ef6d96530cb3791f0dfb16d157220b6
│       ├── a5b0164846456eb4c32115f57cfc3a38
│       ├── bb6d4c2870158394ea6b4c6ec804d61f
│       ├── c3ca6ad75c459ecccfdb10b421cbb489
│       └── ec7e8e8ce16f3fe945d3c4c0955315d3
├── foo
│   ├── bar
│   │   └── tata.txt
│   ├── titi.txt
│   └── toto.txt
└── toto.txt

5 directories, 13 files

EXECUTION DE LA COMMANDE SHELL cat toto.txt
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed non risus. 
Suspendisse lectus tortor, dignissim sit amet, adipiscing nec, ultricies sed, dolor. 
Cras elementum ultrices diam. Maecenas ligula massa, varius a, semper congue, 
euismod non, mi. Proin porttitor, orci nec nonummy molestie, enim est eleifend mi, 
non fermentum diam nisl sit amet erat. Duis semper. 
Duis arcu massa, scelerisque vitae, consequat in, pretium a, enim. 
Pellentesque congue. Ut in risus volutpat libero pharetra tempor. 
Cras vestibulum bibendum augue. Praesent egestas leo in pede. 
Praesent blandit odio eu enim. Pellentesque sed dui ut augue blandit sodales. 
Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; 
Aliquam nibh. Mauris ac mauris sed pede pellentesque fermentum. 
Maecenas adipiscing ante non diam sodales hendrerit.
Ut velit mauris, egestas sed, gravida nec, ornare ut, mi. 
Aenean ut orci vel massa suscipit pulvinar. Nulla sollicitudin. 
Fusce varius, ligula non tempus aliquam, nunc turpis ullamcorper nibh, in tempus 
sapien eros vitae ligula. Pellentesque rhoncus nunc et augue. Integer id felis. 
Curabitur aliquet pellentesque diam. Integer quis metus vitae elit lobortis egestas. 

Aliquam convallis sollicitudin purus. Praesent aliquam, enim at fermentum mollis, ligula massa adipiscing nisl, ac euismod nibh nisl eu lectus. Fusce vulputate sem at sapien. Vivamus leo. Aliquam euismod libero eu enim. Nulla nec felis sed leo placerat imperdiet. Aenean suscipit nulla in justo. Suspendisse cursus rutrum augue. Nulla tincidunt tincidunt mi. Curabitur iaculis, lorem vel rhoncus faucibus, felis magna fermentum augue, et ultricies lacus lorem varius purus. Curabitur eu amet.